### PR TITLE
Detect flexvolume dirpath

### PR DESCRIPTION
**Pull request by LLParse**
https://github.com/rancher/longhorn/issues/266

WIP needs some more testing

### CODE REVIEW

File: app/driver.go [hunk 1]
```diff
@@ -370,6 +374,115 @@ func getFlexvolumeDaemonSetSpec(image, flexvolumeDir string) *appsv1beta2.Daemon
...
+	ctx, cancel := context.WithTimeout(nil, 1*time.Second)
+	s.Shutdown(ctx)
+	cancel()
+
+	// cleanup async
+
+	return flexvolumeDir, nil
+}
+
 func discoverFlexvolumeDir(kubeClient *clientset.Clientset) (dir string, err error) {
```
Comment (LLParse) [inline app/driver.go]: @yasker What should we do with the old method of detection? Just delete it?

Comment (yasker) [reply app/driver.go]: The old way is much faster. Check the version and if it's <= 1.8, run the old way first. If the old way cannot get the result, run the new way.

Comment (yasker) [reply app/driver.go]: In fact, thinking about it again, only 1.8 can use it. Not worth the effort to keep it. Let's remove it and switch altogether.

File: app/driver.go [hunk 2]
```diff
@@ -371,29 +372,134 @@ func getFlexvolumeDaemonSetSpec(image, flexvolumeDir string) *appsv1beta2.Daemon
...
+	// start server to receive results
+	doneChan := make(chan struct{})
+	s := &http.Server{
+		Addr: fmt.Sprintf(":%d", serverPort),
 	}
-	value, ok := jsonParsed.Path("kubeletconfig.volumePluginDir").Data().(string)
-	if !ok {
-		logrus.Infof("cannot find volumePluginDir key in node config, assume it's default")
-		return DefaultFlexvolumeDir, nil
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
```
Comment (yasker) [inline app/driver.go]: Start a new web service for this? I think it's overkill.

I think you can just run a pod to completion then check it's log.

Comment (LLParse) [reply app/driver.go]: I tend to agree, but haven't figured out how to access the logs. I'll dig into it more.

File: app/driver.go [hunk 3]
```diff
@@ -206,15 +205,15 @@ func deployCSIDriver(kubeClient *clientset.Clientset, c *cli.Context, managerIma
...
-		notify := func(kubeClient *clientset.Clientset, f func(*clientset.Clientset)) {
-			f(kubeClient)
-			wg.Done()
-		}
-
-		wg.Add(3)
-		go notify(kubeClient, attacherDeployment.Cleanup)
-		go notify(kubeClient, provisionerDeployment.Cleanup)
-		go notify(kubeClient, pluginDeployment.Cleanup)
+		util.RunAsync(&wg, func() {
```
Comment (LLParse) [inline app/driver.go]: I know this is now an off-topic refactor, but an older revision of the code (that created a webservice and resources) relied on the same logic, so I genericized the function. I can split this into a separate commit (or remove) if need be

File: app/driver_flexvolume.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		logrus.Warnf("Failed to deploy flexvolume detection pod: %v", err)
+	}
+
+	defer func() {
+		if err := kubeClient.CoreV1().Pods(namespace).Delete(DETECT_FLEXVOLUME_NAME, &metav1.DeleteOptions{}); err != nil {
+			logrus.Warnf("Failed to delete flexvolume detection pod: %v", err)
+		}
+	}()
+
+	// FIXME wait until completion
```
Comment (LLParse) [inline app/driver_flexvolume.go]: need to fix

**Comment by LLParse** [comment]
@yasker ready for review

File: app/driver_flexvolume.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+		return "", err
+	}
+
+	defer func() {
+		if err := kubeClient.CoreV1().Pods(namespace).Delete(DetectFlexVolumeName, &metav1.DeleteOptions{}); err != nil {
+			logrus.Warnf("Failed to delete flexvolume detection pod: %v", err)
+		}
+	}()
+
+	for delay := 1 * time.Second; ; delay *= 2 {
```
Comment (yasker) [inline app/driver_flexvolume.go]: Don't wait indefinitely.

Also, the exponentially back off is too long in the second level. Just do a normal 1-second check for a certain period of time is enough.

**Comment by LLParse** [comment]
@yasker

File: app/driver_flexvolume.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+		return "", err
+	}
+
+	defer func() {
+		if err := kubeClient.CoreV1().Pods(namespace).Delete(DetectFlexVolumeName, &metav1.DeleteOptions{}); err != nil {
+			logrus.Warnf("Failed to delete flexvolume detection pod: %v", err)
+		}
+	}()
+
+	for i := 0; i < 120; i++ {
```
Comment (yasker) [inline app/driver_flexvolume.go]: Can you use a constant?

Comment (LLParse) [reply app/driver_flexvolume.go]: Done

**Comment by yasker** [comment]
Merged, thanks!

Please update the doc in another PR.

**Pull request merged**